README.md
## Aewan - multi-layered ascii-art/animation editor

Aewan is a curses-based program that allows for the creation and editing of
ascii art. The native installation packages are customized for integration with
[Asciiville](https://github.com/doctorfree/Asciiville).

The `asciiville-aewan` package gets installed as part of the `Asciiville`
initialization process. See the
[Asciiville README](https://github.com/doctorfree/Asciiville#readme)
for more information.

Many of the [Doctorfree projects](https://github.com/doctorfree) are designed
to integrate with each other including
[Asciiville](https://github.com/doctorfree/Asciiville#readme),
[MirrorCommand](https://github.com/doctorfree/MirrorCommand#readme),
[MusicPlayerPlus](https://github.com/doctorfree/MusicPlayerPlus#readme), and
[RoonCommandLine](https://github.com/doctorfree/RoonCommandLine#readme).

## Table of Contents

1. [Overview](#overview)
1. [Requirements](#requirements)
1. [Installation](#installation)
1. [Removal](#removal)
1. [Quick start](#quick-start)
1. [Asking for help](#asking-for-help)
1. [Building aewan from source](#building-aewan-from-source)
1. [Contributing](#contributing)

## Overview

Aewan is a curses-based program that allows for the creation and
editing of ascii art. The user is able to move the cursor around the screen by
means of the arrow keys and 'paint' characters by pressing the corresponding
keys. There are dialog boxes that allow the user to choose foreground and
background colors, as well as bold and blink attributes. The user may also
select rectangular areas of the canvas in order to move, copy and paste them.
Aewan also supports 'intelligent' horizontal and vertical flipping (e.g.
converts '\' to '/', etc).

What sets Aewan apart from similar projects is the fact that it can work with
multiple layers, and has the ability to turn transparency and visibility on and
off for each layer. A layer dialog is provided through which the user can
change the order of the layers. Thus, each layer can be edited independently in
order to generate a composite drawing. Instead of using the layers for
compositing, it is also possible to use the layers as frames for an animation,
thus enabling the user to create ascii animations with Aewan (note: to be fully
implemented next release).

The file format is easy to parse, so it is easy to write a terminal-based
application that uses the Aewan files to display onscreen. Currently it has
been tested on the Linux terminal, rxvt, xterm, the Cygwin terminal and the
FreeBSD console. Although it is already quite stable (I am already using it on
some projects of mine), it is a little rough on the edges, but that can be
worked out in the near future.

[![License: GPL v2](https://img.shields.io/badge/License-GPLv2-blue.svg)](https://www.gnu.org/licenses/gpl-2.0)

## Requirements

Asciiville-Aewan is compiled and packaged for installation on:

- Arch Linux (x86_64)
- Fedora Linux (x86_64)
- Raspberry Pi OS (armhf)
- Ubuntu Linux (amd64)

## Installation

Asciiville-Aewan v1.0.0 and later can be installed on Linux systems using
the Arch packaging format, the Debian packaging format, or the Red Hat
Package Manager (RPM).

### Supported platforms

Asciiville-Aewan has been tested successfully on the following platforms:

- **Arch Linux 2022.07.01**
    - `asciiville-aewan_<version>-<release>-x86_64.pkg.tar.zst`
- **Ubuntu Linux 20.04**
    - `asciiville-aewan_<version>-<release>.amd64.deb`
- **Fedora Linux 36**
    - `asciiville-aewan_<version>-<release>.x86_64.rpm`
- **Raspberry Pi OS**
    - `asciiville-aewan_<version>-<release>.armhf.deb`

### Debian package installation

Many Linux distributions, most notably Ubuntu and its derivatives, use the
Debian packaging system.

To tell if a Linux system is Debian based it is usually sufficient to
check for the existence of the file `/etc/debian_version` and/or examine the
contents of the file `/etc/os-release`.

To install on a Debian based Linux system, download the latest Debian format
package from the
[asciiville-aewan Releases](https://github.com/doctorfree/asciiville-aewan/releases).

Install the asciiville-aewan package by executing the command

```console
sudo apt install ./asciiville-aewan_<version>-<release>.amd64.deb
```
or
```console
sudo dpkg -i ./asciiville-aewan_<version>-<release>.amd64.deb
```

**NOTE:** In some cases you may see a warning message when installing the
Debian package. The message:

Repository is broken: asciiville-aewan:amd64 (= <version-<release>) has no Size
information can safely be ignored. This is an issue with the Debian packaging
system and has no effect on the installation.

### RPM Package installation

Red Hat Linux, SUSE Linux, and their derivatives use the RPM packaging
format. RPM based Linux distributions include Fedora, AlmaLinux, CentOS,
openSUSE, OpenMandriva, Mandrake Linux, Red Hat Linux, and Oracle Linux.

To install on an RPM based Linux system, download the latest RPM format
package from the
[asciiville-aewan Releases](https://github.com/doctorfree/asciiville-aewan/releases).

Install the asciiville-aewan package by executing the command

```console
sudo yum localinstall ./asciiville-aewan_<version>-<release>.x86_64.rpm
```
or
```console
sudo rpm -i ./asciiville-aewan_<version>-<release>.x86_64.rpm
```

### Arch Package installation

Arch Linux, Manjaro, and other Arch Linux derivatives use the Pacman packaging
format. In addition to Arch Linux, Arch based Linux distributions include
ArchBang, Arch Linux, Artix Linux, ArchLabs, Asahi Linux, BlackArch,
Chakra Linux, EndeavourOS, Frugalware Linux, Garuda Linux,
Hyperbola GNU/Linux-libre, LinHES, Manjaro, Parabola GNU/Linux-libre,
SteamOS, and SystemRescue.

To install on an Arch based Linux system, download the latest Pacman format
package from the
[asciiville-aewan Releases](https://github.com/doctorfree/asciiville-aewan/releases).

Install the asciiville-aewan package by executing the command

```console
sudo pacman -U ./asciiville-aewan_<version>-<release>-x86_64.pkg.tar.zst
```

## Removal

On Debian based Linux systems where the asciiville-aewan package was installed
using the asciiville-aewan Debian format package, remove the asciiville-aewan
package by executing the command:

```console
    sudo apt remove asciiville-aewan
```
or
```console
    sudo dpkg -r asciiville-aewan
```

On RPM based Linux systems where the asciiville-aewan package was installed
using the asciiville-aewan RPM format package, remove the asciiville-aewan
package by executing the command:

```console
    sudo yum remove asciiville-aewan
```
or
```console
    sudo rpm -e asciiville-aewan
```

On Arch based Linux systems where the asciiville-aewan package was installed
using the asciiville-aewan Pacman format package, remove the asciiville-aewan
package by executing the command:

```console
    sudo pacman -Rs asciiville-aewan
```

The asciiville-aewan package can be removed by executing the "Uninstall"
script in the asciiville-aewan source directory:

```console
    git clone https://github.com/doctorfree/asciiville-aewan.git
    cd asciiville-aewan
    ./Uninstall
```

## Building aewan from source

Aewan can be compiled, packaged, and installed from the source code
repository. This should be done as a normal user with `sudo` privileges:

```
# Retrieve the source code from the repository
git clone https://github.com/doctorfree/asciiville-aewan.git
# Enter the asciiville-aewan source directory
cd asciiville-aewan
# Compile the asciiville-aewan components and create an installation package
./mkpkg
# Install asciiville-aewan and its dependencies
./Install
```

These steps are detailed below.

### Clone asciiville-aewan repository

```
git clone https://github.com/doctorfree/asciiville-aewan.git
cd asciiville-aewan
```

**[Note:]** The `mkpkg` script in the top level of the asciiville-aewan
repository can be used to build an installation package on all supported
platforms. After cloning, `cd asciiville-aewan` and `./mkpkg`. The resulting
installation package(s) will be found in `./releases/<version>/`.

### Install packaging dependencies

asciiville-aewan components have packaging dependencies on the following:

On Debian based systems like Ubuntu Linux, install packaging dependencies via:

```
sudo apt install dpkg
```

On RPM based systems like Fedora Linux, install packaging dependencies via:

```
sudo dnf install rpm-build rpm-devel rpmlint rpmdevtools
```

### Build and package asciiville-aewan

To build and package asciiville-aewan, execute the command:

```
./mkpkg
```

On Debian based systems like Ubuntu Linux, the `mkpkg` scripts executes
`scripts/mkdeb.sh`.

On RPM based systems like Fedora Linux, the `mkpkg` scripts executes
`scripts/mkrpm.sh`.

On PKGBUILD based systems like Arch Linux, the `mkpkg` scripts executes
`scripts/mkaur.sh`.

### Install asciiville-aewan from source build

After successfully building and packaging asciiville-aewan with either
`./mkpkg`, install the asciiville-aewan package with the command:

```
./Install
```

## Contributing

There are a variety of ways to contribute to the asciiville-aewan project.
All forms of contribution are appreciated and valuable. Also, it's fun to
collaborate. Here are a few ways to contribute to the further improvement
and evolution of asciiville-aewan:

### Testing and Issue Reporting

asciiville-aewan is fairly complex with many components, features, options,
configurations, and use cases. Although currently only supported on
Linux platforms, there are a plethora of Linux platforms on which
asciiville-aewan can be deployed. Testing all of the above is time consuming
and tedious. If you have a Linux platform on which you can install
asciiville-aewan and you have the time and will to put it through its paces,
then issue reports on problems you encounter would greatly help improve the
robustness and quality of asciiville-aewan. Open issue reports at
[https://github.com/doctorfree/asciiville-aewan/issues](https://github.com/doctorfree/asciiville-aewan/issues)

### Sponsor asciiville-aewan

asciiville-aewan is completely free and open source software. All of the
asciiville-aewan components are freely licensed and may be copied, modified,
and redistributed freely. Nobody gets paid, nobody is making any money,
it's a project fully motivated by curiousity and love of music. However,
it does take some money to procure development and testing resources.
Right now asciiville-aewan needs a multi-boot test platform to extend support
to a wider variety of Linux platforms and potentially Mac OS X.

If you have the means and you would like to sponsor asciiville-aewan development,
testing, platform support, and continued improvement then your monetary
support could play a very critical role. A little bit goes a long way
in asciiville-aewan. For example, a bootable USB SSD device could serve as a 
means of porting and testing support for additional platforms. Or, a
decent cup of coffee could be the difference between a bug filled
release and a glorious musical adventure.

Sponsor the asciiville-aewan project at
[https://github.com/sponsors/doctorfree](https://github.com/sponsors/doctorfree)

### Contribute to Development

If you have programming skills and find the management and ease-of-use of
digital music libraries to be an interesting area, you can contribute to
asciiville-aewan development through the standard Github "fork, clone,
pull request" process. There are many guides to contributing to Github hosted
open source projects on the Internet. A good one is available at
[https://www.dataschool.io/how-to-contribute-on-github/](https://www.dataschool.io/how-to-contribute-on-github/). Another short succinct guide is at
[https://gist.github.com/MarcDiethelm/7303312](https://gist.github.com/MarcDiethelm/7303312).

Once you have forked and cloned the asciiville-aewan repository, it's time to
setup a development environment. Although many of the asciiville-aewan commands
are Bash shell scripts, there are also applicatons written in C and C++ along
with documentation in Markdown format, configuration files in a variety of
formats, examples, screenshots, video demos, build scripts, packaging, and more.

To compile `asciiville-aewan` from source, run the command:

```
./build aewan
```

On Debian (e.g. Ubuntu), PKGBUILD (e.g. Arch) and RPM (e.g. Fedora) based
systems the asciiville-aewan installation package can be created with the
`mkpkg` scripts. The `mkpkg` script determines which platform it is on
and executes the appropriate build and packaging script in the `scripts/`
directory. These scripts invoke the build scripts for each of the projects
included with asciiville-aewan, populate a distribution tree, and call the
respective packaging utilities. Packages are saved in the
`./releases/<version>/` folder. Once a package has been created
it can be installed with the `Install` script.

It's not necessary to have C/C++ expertise to contribute to asciiville-aewan
development. Many of the asciiville-aewan commands are Bash scripts and require
no compilaton. Script commands reside in the `bin` and `share/scripts`
directories. To modify a shell script, install asciiville-aewan and edit the
`bin/<script>` or `share/scripts/<script.sh>` you wish to improve.
After making your changes simply copy the revised script to `/usr/bin`
or `/usr/share/asciiville-aewan/scripts` and test your changes.

Feel free to email me at github@ronrecord.com with questions or comments.



CHANGELOG
1.0.0r1

 * Separate standalone package
 * Integration with Asciiville project
 * C/compiler porting changes for compilation on Arch Linux

1.0.01

 * aecat now accepts "comment" format (which extracts metadata from the file)
   and -L option (which outputs layer count) - contributed by 
   Gerfried Fuchs (alfie@ist.org)
 * corrected aeff (aewan file format) escape sequence interpretation bug
   that was causing special characters in metainfo to get garbled
 * added "resize layer" funcionality, also contributed by 
   Gerfried Fuchs (alfie@ist.org)
 * both 'Load file' and 'Save file' now use dialog boxes that let the
   user browse directories 
 * added a new export utility: aemakeflic, which converts aewan documents
   into animations to be played in the 'less' command or as a shell script
 * added a confirmation question when quitting the program or clearing
   the document to prevent against accidental loss of current work
 * in the version number, the revision will be conventionally written
   with two digits from now on

0.9.7 

 * cleaned up Makefile.in to make the bores/ directory (remnants of a library
   that used to be separate from aewan) compile together with the rest
   of the project instead of separately as libbores.a

0.9.6 - release date: 2005-02-19 (minor bugfix release)

 * fixed segfault bug which resulted from trying to correct cursor coords
   when a layer had just been added (reported in Debian bug tracking system)

0.9.5 - release date: 2005-02-18

 * menu introduced in order to make the interface more friendly
 * "new file" (i.e. clear everything and start from scratch) funcionality added
 * added patches from Gerfried Fuchs (alfie@ist.org):
      - aecat now produces more XHTML-compliant html and supports
        background color
      - user can now manipulate upper-left corner of selection in select mode


pkg/mkaur.sh
#!/bin/bash
#
# mkaur.sh - front end for Arch Linux PKGBUILD distribution creation

PKG="asciiville-aewan"
PKG_NAME="asciiville-aewan"
FULLNAME="Ronald Record"
EMAIL="ronaldrecord@gmail.com"
DESTDIR="usr"
ARCH=x86_64
SUDO=sudo
GCI=

have_makepkg=`type -p makepkg`
[ "${have_makepkg}" ] || {
  echo "Arch Linux packaging tools do not appear to be installed"
  echo "Are you on the appropriate Linux system with packaging requirements ?"
  echo "Exiting"
  exit 1
}

arch_arch=`uname -m`
[ "${arch_arch}" == "${ARCH}" ] || ARCH=${arch_arch}

if [ "${__ASC_SRC__}" ]
then
  SRC="${__ASC_SRC__}"
else
  SRC="${HOME}/src/${PKG_NAME}"
fi

[ -f "${SRC}/VERSION" ] || {
  [ -f "/builds/doctorfree/${PKG_NAME}/VERSION" ] || {
    echo "${SRC}/VERSION does not exist. Exiting."
    exit 1
  }
  SRC="/builds/doctorfree/${PKG_NAME}"
  GCI=1
# SUDO=
}

. "${SRC}/VERSION"
PKG_VER=${VERSION}
PKG_REL=${RELEASE}

umask 0022

# Subdirectory in which to create the distribution files
OUT_DIR="dist/${PKG_NAME}_${PKG_VER}"

[ -d "${SRC}" ] || {
    echo "$SRC does not exist or is not a directory. Exiting."
    exit 1
}

cd "${SRC}"

${SUDO} rm -rf ${PKG}
mkdir ${PKG}
cp VERSION ${PKG}/VERSION
cp pkg/aur/PKGBUILD ${PKG}/PKGBUILD
cp pkg/aur/${PKG}.install ${PKG}/${PKG}.install
cp pkg/aur/.SRCINFO ${PKG}/.SRCINFO
cp pkg/aur/makepkg.conf ${PKG}/makepkg.conf

echo "Creating ${PKG_NAME} PKGBUILD distribution archive"
tar cf - ${PKG} | \
    gzip -9 > ${PKG_NAME}-pkgbuild-${PKG_VER}-${PKG_REL}.tar.gz
mv ${PKG_NAME}-pkgbuild-${PKG_VER}-${PKG_REL}.tar.gz ${PKG}

echo "Building ${PKG_NAME}_${PKG_VER} AUR package"
cd "${SRC}/${PKG}"
makepkg --force --log --cleanbuild --noconfirm --syncdeps

# Rename package if necessary
for zstfile in *.zst
do
  [ "${zstfile}" == "*.zst" ] && continue
  newnam=`echo ${zstfile} | sed -e "s/${PKG}-${PKG_VER}-${PKG_REL}/${PKG_NAME}_${PKG_VER}-${PKG_REL}/"`
  [ "${zstfile}" == "${newnam}" ] && continue
  mv ${zstfile} ${newnam}
done

[ "${GCI}" ] || {
  [ -d ../releases ] || mkdir ../releases
  [ -d ../releases/${PKG_VER} ] || mkdir ../releases/${PKG_VER}
  ${SUDO} cp *.zst *.tar.gz ../releases/${PKG_VER}
}


pkg/mkrpm.sh
#!/bin/bash
PKG="asciiville-aewan"
SRC_NAME="asciiville-aewan"
PKG_NAME="asciiville-aewan"
DESTDIR="usr"
SRC=${HOME}/src
SUDO=sudo
GCI=

[ -f "${SRC}/${SRC_NAME}/VERSION" ] || {
  [ -f "/builds/doctorfree/${SRC_NAME}/VERSION" ] || {
    echo "$SRC/$SRC_NAME/VERSION does not exist. Exiting."
    exit 1
  }
  SRC="/builds/doctorfree"
  SUDO=
  GCI=1
}

. "${SRC}/${SRC_NAME}/VERSION"
PKG_VER=${VERSION}
PKG_REL=${RELEASE}

umask 0022

# Subdirectory in which to create the distribution files
OUT_DIR="dist/${PKG_NAME}_${PKG_VER}"

[ -d "${SRC}/${SRC_NAME}" ] || {
    echo "$SRC/$SRC_NAME does not exist or is not a directory. Exiting."
    exit 1
}

cd "${SRC}/${SRC_NAME}"

# Install required development environment tools
PKGS="automake libtool ncurses-devel"
${SUDO} dnf -y groupinstall "Development Tools" "Development Libraries"
${SUDO} dnf -y install ${PKGS} pandoc zip

# Build aewan
if [ -x build ]
then
  ./build
else
  [ -f aewan ] || {
    autoreconf -vi
    ./configure --prefix=/usr
    make
  }
fi
chmod +x aewan aecat aemakeflic

${SUDO} rm -rf dist
mkdir dist

[ -d ${OUT_DIR} ] && rm -rf ${OUT_DIR}
mkdir ${OUT_DIR}

for dir in "${DESTDIR}" "${DESTDIR}/bin" "${DESTDIR}/share" \
           "${DESTDIR}/share/doc" "${DESTDIR}/share/doc/${PKG}" \
           "${DESTDIR}/share/man" "${DESTDIR}/share/man/man1" \
           "${DESTDIR}/share/man/man5" "${DESTDIR}/share/${PKG}"
do
    [ -d ${OUT_DIR}/${dir} ] || ${SUDO} mkdir ${OUT_DIR}/${dir}
    ${SUDO} chown root:root ${OUT_DIR}/${dir}
done

# Install aewan
${SUDO} cp aewan aecat aemakeflic ${OUT_DIR}/${DESTDIR}/bin
${SUDO} cp man/man1/*.1 ${OUT_DIR}/${DESTDIR}/share/man/man1
${SUDO} cp man/man5/*.5 ${OUT_DIR}/${DESTDIR}/share/man/man5

${SUDO} cp aewan-README ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp copyright ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp CHANGELOG ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp COPYING ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp README.md ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp LICENSE ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp TODO ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp VERSION ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} pandoc -f gfm README.md | ${SUDO} tee ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/README.html > /dev/null
${SUDO} gzip -9 ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/CHANGELOG

${SUDO} chmod 755 ${OUT_DIR}/${DESTDIR}/bin/* ${OUT_DIR}/${DESTDIR}/bin
${SUDO} chown -R root:root ${OUT_DIR}/${DESTDIR}

echo "Building ${PKG_NAME}_${PKG_VER} rpm package"

[ -d pkg/rpm ] && cp -a pkg/rpm ${OUT_DIR}/rpm
[ -d ${OUT_DIR}/rpm ] || mkdir ${OUT_DIR}/rpm

have_rpm=`type -p rpmbuild`
[ "${have_rpm}" ] || {
  have_yum=`type -p yum`
  if [ "${have_yum}" ]
  then
    ${SUDO} yum -y install rpm-build
  else
    ${SUDO} apt-get update
    export DEBIAN_FRONTEND=noninteractive
    ${SUDO} ln -fs /usr/share/zoneinfo/America/Los_Angeles /etc/localtime
    ${SUDO} apt-get install rpm -y
    ${SUDO} dpkg-reconfigure --frontend noninteractive tzdata
  fi
}

rpmbuild -ba --build-in-place \
   --define "_topdir ${OUT_DIR}" \
   --define "_sourcedir ${OUT_DIR}" \
   --define "_version ${PKG_VER}" \
   --define "_release ${PKG_REL}" \
   --buildroot ${SRC}/${SRC_NAME}/${OUT_DIR}/BUILDROOT \
   ${OUT_DIR}/rpm/${PKG_NAME}.spec

# Rename RPMs if necessary
for rpmfile in ${OUT_DIR}/RPMS/*/*.rpm
do
  [ "${rpmfile}" == "${OUT_DIR}/RPMS/*/*.rpm" ] && continue
  rpmbas=`basename ${rpmfile}`
  rpmdir=`dirname ${rpmfile}`
  newnam=`echo ${rpmbas} | sed -e "s/${PKG_NAME}-${PKG_VER}-${PKG_REL}/${PKG_NAME}_${PKG_VER}-${PKG_REL}/"`
  [ "${rpmbas}" == "${newnam}" ] && continue
  mv ${rpmdir}/${rpmbas} ${rpmdir}/${newnam}
done

${SUDO} cp ${OUT_DIR}/RPMS/*/*.rpm dist

[ "${GCI}" ] || {
    [ -d releases ] || mkdir releases
    [ -d releases/${PKG_VER} ] || mkdir releases/${PKG_VER}
    ${SUDO} cp ${OUT_DIR}/RPMS/*/*.rpm releases/${PKG_VER}
}


pkg/build-aewan.sh
#!/bin/bash
#

usage() {
  printf "\nUsage: ./build-aewan.sh [-Ci] [-d destdir] [-p prefix] [-t] [-u]"
  printf "\nWhere:"
  printf "\n\t-C indicates run configure and exit"
  printf "\n\t-d destdir specifies installation destination root (default /)"
  printf "\n\t-i indicates configure, build, and install"
  printf "\n\t-p prefix specifies installation prefix (default /usr)"
  printf "\n\t-t indicates configure with tensorflow support"
  printf "\n\t-u displays this usage message and exits\n"
  printf "\nNo arguments: configure, build\n"
  exit 1
}

install_aewan() {
  [ -d ${PREFIX}/usr ] || sudo mkdir -p ${PREFIX}/usr
  [ -d ${PREFIX}/usr/bin ] || sudo mkdir -p ${PREFIX}/usr/bin
  sudo cp aewan aecat aemakeflic ${PREFIX}/usr/bin
  [ -d ${PREFIX}/usr/share ] || sudo mkdir -p ${PREFIX}/usr/share
  [ -d ${PREFIX}/usr/share/man ] || sudo mkdir -p ${PREFIX}/usr/share/man
  [ -d ${PREFIX}/usr/share/man/man1 ] || sudo mkdir -p ${PREFIX}/usr/share/man/man1
  [ -d ${PREFIX}/usr/share/man/man5 ] || sudo mkdir -p ${PREFIX}/usr/share/man/man5
  sudo cp man/man1/*.1 ${PREFIX}/usr/share/man/man1
  sudo cp man/man5/*.5 ${PREFIX}/usr/share/man/man5
}

CONFIGURE_ONLY=
INSTALL=
PREFIX=
DESTDIR=
TENSOR=
while getopts "Cd:ip:tu" flag; do
    case $flag in
        C)
            CONFIGURE_ONLY=1
            ;;
        d)
            DESTDIR="${OPTARG}"
            ;;
        i)
            INSTALL=1
            ;;
        p)
            PREFIX="${OPTARG}"
            ;;
        t)
            TENSOR=1
            ;;
        u)
            usage
            ;;
    esac
done
shift $(( OPTIND - 1 ))

destdir=
prefix="--prefix=/usr"
tensor=
[ "${DESTDIR}" ] && destdir="--destdir=${DESTDIR}"
[ "${PREFIX}" ] && prefix="--prefix=${PREFIX}"
[ "${TENSOR}" ] && tensor="--with-tensorflow"

[ -x aewan ] && {
  if [ "${INSTALL}" ]
  then
    install_aewan
  fi
  exit 0
}

prefix="--prefix=/usr"
[ "${PREFIX}" ] && prefix="--prefix=${PREFIX}"
autoreconf -vi
./configure ${prefix}

[ "${CONFIGURE_ONLY}" ] && exit 0

make
chmod +x aewan aecat aemakeflic

if [ "${INSTALL}" ]
then
  install_aewan
fi



pkg/mkdeb.sh
#!/bin/bash
PKG="asciiville-aewan"
SRC_NAME="asciiville-aewan"
PKG_NAME="asciiville-aewan"
DEBFULLNAME="Ronald Record"
DEBEMAIL="ronaldrecord@gmail.com"
DESTDIR="usr"
SRC=${HOME}/src
ARCH=amd64
SUDO=sudo
GCI=

dpkg=`type -p dpkg-deb`
[ "${dpkg}" ] || {
    echo "Debian packaging tools do not appear to be installed on this system"
    echo "Are you on the appropriate Linux system with packaging requirements ?"
    echo "Exiting"
    exit 1
}
dpkg_arch=`dpkg --print-architecture`
[ "${dpkg_arch}" == "${ARCH}" ] || ARCH=${dpkg_arch}

[ -f "${SRC}/${SRC_NAME}/VERSION" ] || {
  [ -f "/builds/doctorfree/${SRC_NAME}/VERSION" ] || {
    echo "$SRC/$SRC_NAME/VERSION does not exist. Exiting."
    exit 1
  }
  SRC="/builds/doctorfree"
  GCI=1
# SUDO=
}

. "${SRC}/${SRC_NAME}/VERSION"
PKG_VER=${VERSION}
PKG_REL=${RELEASE}

umask 0022

# Subdirectory in which to create the distribution files
OUT_DIR="dist/${PKG_NAME}_${PKG_VER}"

[ -d "${SRC}/${SRC_NAME}" ] || {
    echo "$SRC/$SRC_NAME does not exist or is not a directory. Exiting."
    exit 1
}

cd "${SRC}/${SRC_NAME}"

# Install required development environment tools
${SUDO} apt install autotools-dev autoconf libtool build-essential libncurses-dev

# Build aewan
if [ -x build ]
then
  ./build
else
  [ -f aewan ] || {
    autoreconf -vi
    ./configure --prefix=/usr
    make
  }
fi
chmod +x aewan aecat aemakeflic

${SUDO} rm -rf dist
mkdir dist

[ -d ${OUT_DIR} ] && rm -rf ${OUT_DIR}
mkdir ${OUT_DIR}
mkdir ${OUT_DIR}/DEBIAN
chmod 755 ${OUT_DIR} ${OUT_DIR}/DEBIAN

echo "Package: ${PKG}
Version: ${PKG_VER}-${PKG_REL}
Section: misc
Priority: optional
Architecture: ${ARCH}
Depends: libncursesw6 (>= 6), zlib1g
Maintainer: ${DEBFULLNAME} <${DEBEMAIL}>
Build-Depends: debhelper (>= 11)
Homepage: https://github.com/doctorfree/asciiville-aewan
Description: Aewan is a curses-based program for the creation and editing of ascii art" > ${OUT_DIR}/DEBIAN/control

chmod 644 ${OUT_DIR}/DEBIAN/control

for dir in "${DESTDIR}" "${DESTDIR}/bin" "${DESTDIR}/share" \
           "${DESTDIR}/share/doc" "${DESTDIR}/share/man" \
           "${DESTDIR}/share/man/man1" "${DESTDIR}/share/man/man5" \
           "${DESTDIR}/share/doc/${PKG}" "${DESTDIR}/share/${PKG}"
do
    [ -d ${OUT_DIR}/${dir} ] || ${SUDO} mkdir ${OUT_DIR}/${dir}
    ${SUDO} chown root:root ${OUT_DIR}/${dir}
done

# Install aewan
${SUDO} cp aewan aecat aemakeflic ${OUT_DIR}/${DESTDIR}/bin
${SUDO} cp man/man1/*.1 ${OUT_DIR}/${DESTDIR}/share/man/man1
${SUDO} cp man/man5/*.5 ${OUT_DIR}/${DESTDIR}/share/man/man5

${SUDO} cp aewan-README ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp copyright ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp CHANGELOG ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp COPYING ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp README.md ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp LICENSE ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp TODO ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp VERSION ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} pandoc -f gfm README.md | ${SUDO} tee ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/README.html > /dev/null
${SUDO} gzip -9 ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/CHANGELOG

${SUDO} chmod 755 ${OUT_DIR}/${DESTDIR}/bin/* ${OUT_DIR}/${DESTDIR}/bin
${SUDO} chown -R root:root ${OUT_DIR}/${DESTDIR}

cd dist
echo "Building ${PKG_NAME}_${PKG_VER} Debian package"
${SUDO} dpkg --build ${PKG_NAME}_${PKG_VER} ${PKG_NAME}_${PKG_VER}-${PKG_REL}.${ARCH}.deb
cd ${PKG_NAME}_${PKG_VER}
echo "Creating compressed tar archive of ${PKG_NAME} ${PKG_VER} distribution"
tar cf - usr | gzip -9 > ../${PKG_NAME}_${PKG_VER}-${PKG_REL}.tgz

have_zip=`type -p zip`
[ "${have_zip}" ] || {
  ${SUDO} apt-get update
  ${SUDO} apt-get install zip -y
}
echo "Creating zip archive of ${PKG_NAME} ${PKG_VER} distribution"
zip -q -r ../${PKG_NAME}_${PKG_VER}-${PKG_REL}.zip usr

cd ..
[ "${GCI}" ] || {
    [ -d ../releases ] || mkdir ../releases
    [ -d ../releases/${PKG_VER} ] || mkdir ../releases/${PKG_VER}
    ${SUDO} cp *.deb *.tgz *.zip ../releases/${PKG_VER}
}
